sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include <tbb/tbb.h>
+#include <chrono>
+#include <fstream>
+#include <filesystem>
+#include <tiffio.h>
+#include <spdlog/spdlog.h>
+#include "disk_io.h"
+#include "sophiread_core.h"
+
+namespace sophiread {
+
+/**
+ * @brief Timed read raw data to char vector.
+ *
+ * @param[in] in_tpx3
+ * @return std::vector<char>
+ */
+std::vector<char> timedReadDataToCharVec(const std::string &in_tpx3) {
+  auto start = std::chrono::high_resolution_clock::now();
+  auto raw_data = readTPX3RawToCharVec(in_tpx3);
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Read raw data: {} s", elapsed / 1e6);
+
+  return raw_data;
+}
+
+/**
+ * @brief Timed find TPX3H.
+ *
+ * @param[in] rawdata
+ * @return std::vector<TPX3>
+ */
+std::vector<TPX3> timedFindTPX3H(const std::vector<char> &rawdata) {
+  auto start = std::chrono::high_resolution_clock::now();
+  auto batches = findTPX3H(rawdata);
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Locate all headers: {} s", elapsed / 1e6);
+
+  return batches;
+}
+
+/**
+ * @brief Timed locate timestamp.
+ *
+ * @param[in, out] batches
+ * @param[in] rawdata
+ */
+void timedLocateTimeStamp(std::vector<TPX3> &batches, const std::vector<char> &rawdata) {
+  auto start = std::chrono::high_resolution_clock::now();
+  unsigned long tdc_timestamp = 0;
+  unsigned long long gdc_timestamp = 0;
+  unsigned long timer_lsb32 = 0;
+  for (auto &tpx3 : batches) {
+    updateTimestamp(tpx3, rawdata, tdc_timestamp, gdc_timestamp, timer_lsb32);
+  }
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Locate all timestamps: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed hits extraction and clustering via multi-threading.
+ *
+ * @param[in, out] batches
+ * @param[in] rawdata
+ * @param[in] config
+ */
+void timedProcessing(std::vector<TPX3> &batches, const std::vector<char> &raw_data, const IConfig &config) {
+  auto start = std::chrono::high_resolution_clock::now();
+  tbb::parallel_for(tbb::blocked_range<size_t>(0, batches.size()), [&](const tbb::blocked_range<size_t> &r) {
+    // Define ABS algorithm with user-defined parameters for each thread
+    auto abs_alg_mt =
+        std::make_unique<ABS>(config.getABSRadius(), config.getABSMinClusterSize(), config.getABSSpiderTimeRange());
+
+    for (size_t i = r.begin(); i != r.end(); ++i) {
+      auto &tpx3 = batches[i];
+      extractHits(tpx3, raw_data);
+
+      abs_alg_mt->reset();
+      abs_alg_mt->set_method("centroid");
```

Comment:
the method seems like something that would be more efficient to set in the constructor